doc/tunnel/SAI-Proposal-ImprovedP2PTunnels.md

```diff
@@ -0,0 +1,61 @@
+# Introduction
+As of SAI version 1.5 the “tunnel” has a p2mp connotation. It holds the VTEP SIP whereas there is no DIP. The DIP is specified as part of the FDB entry or as part of Next Hop entry.
+
+It is proposed to add the DIP as part of the `sai_tunnel_attr_t` structure as an optional parameter.
+
+The following are the motivations for introducing a DIP in the tunnel structure and to model it as a p2p entity.
+
+- Support Head End Replication as the method for handling BUM traffic.
+- Support per remote IP and per remote IP+VNI Tx and Rx counters.
+- Support the notion of operational status per remote IP.
+- Support flushing FDB per remote IP.
+- Support learning enable/disable per tunnel.
```

Comment:
For these items, a P2P tunnel can be created with the current API as well. By creating multiple P2P tunnels with the same SRC_IP, same functionality will be achieved.

Reply:
Could you please elaborate with examples ?

Reply:
Please see below
```
sai_attribute_t attr;
vector<sai_attribute_t> attrs;

/*****************************************************
 * Create tunnel for each endpoint
 *****************************************************/

sai_object_id_t tunnel2, tunnel3, tunnel4;

attr.id = SAI_TUNNEL_ATTR_TYPE;
attr.value.s32 = SAI_TUNNEL_TYPE_VXLAN;
attrs.push_back(attr);

attr.id = SAI_TUNNEL_ATTR_ENCAP_SRC_IP;
attr.value.ipaddr = 1.1.1.1;
attrs.push_back(attr);

sai_tunnel_api->create_tunnel(&tunnel2, gSwitchId, attrs.size(), attrs.data());
sai_tunnel_api->create_tunnel(&tunnel3, gSwitchId, attrs.size(), attrs.data());
sai_tunnel_api->create_tunnel(&tunnel4, gSwitchId, attrs.size(), attrs.data());

/*****************************************************
 * Create isolation group
 *****************************************************/

sai_object_id_t isolation_group;
 
attr.id = SAI_ISOLATION_GROUP_ATTR_TYPE;
attr.value.s32 = SAI_ISOLATION_GROUP_TYPE_BRIDGE_PORT;
attrs.push_back(attr);

sai_isolation_api->create_isolation_group(&isolation_group, gSwitchId, attrs.size(), attrs.data());

/*****************************************************
 * Create bridge ports for each tunnel
 *****************************************************/
 
sai_object_id_t bridge_port2, bridge_port3, bridge_port4;
 
attr.id = SAI_BRIDGE_PORT_ATTR_TYPE;
attr.value.s32 = SAI_BRIDGE_PORT_TYPE_TUNNEL;
attrs.push_back(attr);

attr.id = SAI_BRIDGE_PORT_ATTR_ADMIN_STATE;
attr.value.booldata = true;
bpt_attrs.push_back(attr);

attr.id = SAI_BRIDGE_PORT_ATTR_FDB_LEARNING_MODE;
attr.value.s32 = SAI_BRIDGE_PORT_FDB_LEARNING_MODE_ENABLE;
attrs.push_back(attr);

attr.id = SAI_BRIDGE_PORT_ATTR_ISOLATION_GROUP;
attr.value.oid = isolation_group;
attrs.push_back(attr);

attr.id = SAI_BRIDGE_PORT_ATTR_TUNNEL_ID;
attr.value.oid = tunnel2;
attrs.push_back(attr);

sai_bridge_api->create_bridge_port(&bridge_port2, gSwitchId, attrs.size(), attrs.data());

attrs.pop_back()
attr.id = SAI_BRIDGE_PORT_ATTR_TUNNEL_ID;
attr.value.oid = tunnel3;
attrs.push_back(attr);

sai_bridge_api->create_bridge_port(&bridge_port3, gSwitchId, attrs.size(), attrs.data());

attrs.pop_back()
attr.id = SAI_BRIDGE_PORT_ATTR_TUNNEL_ID;
attr.value.oid = tunnel4;
attrs.push_back(attr);

sai_bridge_api->create_bridge_port(&bridge_port4, gSwitchId, attrs.size(), attrs.data());

/*****************************************************
 * Add bridge ports to isolation_group
 *****************************************************/
 
sai_object_id_t isolation_group_member2, isolation_group_member3, isolation_group_member4;
 
attr.id = SAI_ISOLATION_GROUP_MEMBER_ATTR_GROUP_ID;
attr.value.oid = isolation_group;
attrs.push_back(attr);

attr.id = SAI_ISOLATION_GROUP_MEMBER_ATTR_ISOLATION_OBJECT;
attr.value.oid = bridge_port2;
attrs.push_back(attr);

sai_isolation_api->create_isolation_group_member(&isolation_group_member2, gSwitchId, attrs.size(), attrs.data());

attrs.pop_back()
attr.id = SAI_ISOLATION_GROUP_MEMBER_ATTR_ISOLATION_OBJECT;
attr.value.oid = bridge_port3;
attrs.push_back(attr);

sai_isolation_api->create_isolation_group_member(&isolation_group_member3, gSwitchId, attrs.size(), attrs.data());

attrs.pop_back()
attr.id = SAI_ISOLATION_GROUP_MEMBER_ATTR_ISOLATION_OBJECT;
attr.value.oid = bridge_port4;
attrs.push_back(attr);

sai_isolation_api->create_isolation_group_member(&isolation_group_member4, gSwitchId, attrs.size(), attrs.data());

/*****************************************************
 * Get counters for a remote IP
 *****************************************************/
 
sai_tunnel_stat_t stat_ids[] = { SAI_TUNNEL_STAT_IN_PACKETS, SAI_TUNNEL_STAT_OUT_PACKETS }
uint64_t stats[2];

sai_tunnel_api->get_tunnel_stats(tunnel2, 2, stat_ids, stats);
```

Reply:
The get attribute call will return an oid which is not created by an application. In the case of SONiC this will mean that there has to be a VID created in syncd as part of the get call specifically for this get attribute call. This just complicates the application in addition to book keeping associated with the L2MC group. 
As has been elaborated in the requirements it is clear that the L2 VXLAN applications will need to go with a P2P model as it is a superset and handles all the requirements.  It would be better to use P2MP model for L3 and P2P instead of allowing for both the models. 
The need to use an L2MC group to specify the flood vector is driven by the fact that we support a P2MP model for L2VXLAN.  If we agree upon using the P2P model for L2VXLAN then the need to specify the multicast flood vector no longer arises. 

The changes suggested in this PR and the use of VLAN_MEMBER_PORT for dot1q and bridge port for dot1d bridges will suffice to specify the flood vector. 

If there is concern in adding DESTIP and P2P/P2MP flags to the tunnel object as the tunnel terminator has these already then allowing for a new bridge port type of type tunnel_terminator would also achieve the goal of specifying the flood vector without having to explicitly use the L2MC group.